# Data Wrangling
The data was downloaded onto an excel table. I imported the excel table to IBM Watson and I downloaded the required libraries to start cleaning up the data.

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [49]:
# The code was removed by Watson Studio for sharing.

Postcode           Borough     Neighbourhood
0      M1A               NaN               NaN
1      M2A               NaN               NaN
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

Let us examine the missing data

In [6]:
missing_data = df.isnull()
missing_data.head(3)

Postcode  Borough  Neighbourhood
0     False     True           True
1     False     True           True
2     False    False          False

In [7]:
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")

Postcode
False    288
Name: Postcode, dtype: int64

Borough
False    211
True      77
Name: Borough, dtype: int64

Neighbourhood
False    210
True      78
Name: Neighbourhood, dtype: int64



As we can see, there are 77 true values, indicating missing data, and per prompt requirement, these will be dropped from the Borough column

In [8]:
df.dropna(subset=["Borough"], axis=0, inplace=True)

#reset the index

df.reset_index(drop=True, inplace=True)

df.head(12)

Postcode           Borough     Neighbourhood
0       M3A        North York         Parkwoods
1       M4A        North York  Victoria Village
2       M5A  Downtown Toronto      Harbourfront
3       M5A  Downtown Toronto       Regent Park
4       M6A        North York  Lawrence Heights
5       M6A        North York    Lawrence Manor
6       M7A      Queen's Park               NaN
7       M9A         Etobicoke  Islington Avenue
8       M1B       Scarborough             Rouge
9       M1B       Scarborough           Malvern
10      M3B        North York   Don Mills North
11      M4B         East York  Woodbine Gardens

We then assign missing neighborhoods to name of corresponding borough

In [9]:
df['Neighbourhood'].replace(np.nan, df['Borough'], inplace=True)
df.head(12)

Postcode           Borough     Neighbourhood
0       M3A        North York         Parkwoods
1       M4A        North York  Victoria Village
2       M5A  Downtown Toronto      Harbourfront
3       M5A  Downtown Toronto       Regent Park
4       M6A        North York  Lawrence Heights
5       M6A        North York    Lawrence Manor
6       M7A      Queen's Park      Queen's Park
7       M9A         Etobicoke  Islington Avenue
8       M1B       Scarborough             Rouge
9       M1B       Scarborough           Malvern
10      M3B        North York   Don Mills North
11      M4B         East York  Woodbine Gardens

Then concatenate by Postcode

In [10]:
df_1= df.groupby('Postcode').agg(lambda x: ','.join(x))

In [11]:
df_2=df_1.reset_index()

Within each Borough, there are multiple Postcodes and so we clean up the data frame to remove any repeats, so that each line has only one Postcode, one Borough, and all the Neighborhoods in that Borough and Postcode.

In [12]:
df_2['Borough']= df_2['Borough'].str.replace('[{}\s]','').str.split(',').apply(set).str.join(',').str.strip(',').str.replace(",{2,}",",")

In [14]:
df_2.head(12)

Postcode      Borough                                      Neighbourhood
0       M1B  Scarborough                                      Rouge,Malvern
1       M1C  Scarborough               Highland Creek,Rouge Hill,Port Union
2       M1E  Scarborough                    Guildwood,Morningside,West Hill
3       M1G  Scarborough                                             Woburn
4       M1H  Scarborough                                          Cedarbrae
5       M1J  Scarborough                                Scarborough Village
6       M1K  Scarborough          East Birchmount Park,Ionview,Kennedy Park
7       M1L  Scarborough                      Clairlea,Golden Mile,Oakridge
8       M1M  Scarborough      Cliffcrest,Cliffside,Scarborough Village West
9       M1N  Scarborough                         Birch Cliff,Cliffside West
10      M1P  Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...
11      M1R  Scarborough                                   Maryvale,Wexford

In [13]:
df_2.shape

(103, 3)

The data frame has 3 rows and 103 columns

# Geocoding
Using the CSV data to merge it to the cleaned table from the previous section. First examine what the data frame looks like

In [14]:
filepath = "https://cocl.us/Geospatial_data"
df_3 = pd.read_csv('https://cocl.us/Geospatial_data')
df_3.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Rename the field "Postal Code" to Postcode to match the previous section, and merge the two data sets to get the required data frame.

In [15]:
df_3.rename(columns={'Postal Code': 'Postcode'}, inplace=True)
df_4 = pd.merge(df_3, df_2, how='inner', on=None, left_on=None, right_on=None,
         left_index=False, right_index=False, sort=True,
         suffixes=('_x', '_y'), copy=True, indicator=False,
         validate=None)

df_4.head()

Postcode   Latitude  Longitude      Borough  \
0      M1B  43.806686 -79.194353  Scarborough   
1      M1C  43.784535 -79.160497  Scarborough   
2      M1E  43.763573 -79.188711  Scarborough   
3      M1G  43.770992 -79.216917  Scarborough   
4      M1H  43.773136 -79.239476  Scarborough   

                          Neighbourhood  
0                         Rouge,Malvern  
1  Highland Creek,Rouge Hill,Port Union  
2       Guildwood,Morningside,West Hill  
3                                Woburn  
4                             Cedarbrae

Then we fix the column order to have Latitude and Longitude as the last two columns, then assign them to the data frame.

In [16]:
column_order = ['Postcode',
 'Borough',
 'Neighbourhood',
 'Latitude',
 'Longitude']
df_5=df_4[column_order]
df_5.head()

Postcode      Borough                         Neighbourhood   Latitude  \
0      M1B  Scarborough                         Rouge,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2      M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3      M1G  Scarborough                                Woburn  43.770992   
4      M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

# Visualization and Clustering
First we import some required libraries

In [1]:
!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim
import requests
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
!conda install -c conda-forge folium=0.5.0 --yes
import folium
print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2019.9.11          |           py36_0         147 KB  conda-forge
    openssl-1.1.1c             |       h516909a_0         2.1 MB  conda-forge
    geopy-1.20.0               |             py_0          57 KB  conda-forge
    ca-certificates-2019.9.11  |       hecc5488_0         144 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0         conda-forge
    geopy:           1.20.0-py_0       conda-forge

The following packages will be UPDATED:

    ca-

Then we look for the coordinates of Toronto

In [2]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="TO_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto, Ontario are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto, Ontario are 43.653963, -79.387207.


Now we show the map of Toronto with the neighborhoods as markers

In [18]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

for lat, lng, borough, neighbourhood in zip(df_5['Latitude'], df_5['Longitude'], df_5['Borough'], df_5['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=4,
        popup=label,
        color='magenta',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

# Using Foursquare API to explore the neighborhoods

In [52]:
# The code was removed by Watson Studio for sharing.

Define a function to get the category

In [26]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Using the get_category_type function, we clean up the json and turn it into a pandas DF. Before we start we need to import certain libraries.

In [27]:
import json
from pandas.io.json import json_normalize 

In [28]:
venues = results['response']['groups'][0]['items']
   
nearby_venues = json_normalize(venues) # flatten JSON

filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name           categories        lat        lng
0    Downtown Toronto         Neighborhood  43.653232 -79.385296
1             Japango     Sushi Restaurant  43.655268 -79.385165
2         Cafe Plenty                 Café  43.654571 -79.389450
3  Sansotei Ramen 三草亭     Ramen Restaurant  43.655157 -79.386501
4          Rolltation  Japanese Restaurant  43.654918 -79.387424

Now we explore the nearby venues!

In [29]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
      
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius)
            
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [31]:
toronto_venues = getNearbyVenues(names=df_5['Neighbourhood'],
                                   latitudes=df_5['Latitude'],
                                   longitudes=df_5['Longitude']
                                  )

Rouge,Malvern
Highland Creek,Rouge Hill,Port Union
Guildwood,Morningside,West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park,Ionview,Kennedy Park
Clairlea,Golden Mile,Oakridge
Cliffcrest,Cliffside,Scarborough Village West
Birch Cliff,Cliffside West
Dorset Park,Scarborough Town Centre,Wexford Heights
Maryvale,Wexford
Agincourt
Clarks Corners,Sullivan,Tam O'Shanter
Agincourt North,L'Amoreaux East,Milliken,Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview,Henry Farm,Oriole
Bayview Village
Silver Hills,York Mills
Newtonbrook,Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park,Don Mills South
Bathurst Manor,Downsview North,Wilson Heights
Northwood Park,York University
CFB Toronto,Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens,Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West,Riverdale
The Beaches West,Indi

In [32]:
print(toronto_venues.shape)
toronto_venues.head()

(1333, 7)


Neighborhood  Neighborhood Latitude  \
0                         Rouge,Malvern              43.806686   
1  Highland Creek,Rouge Hill,Port Union              43.784535   
2       Guildwood,Morningside,West Hill              43.763573   
3       Guildwood,Morningside,West Hill              43.763573   
4       Guildwood,Morningside,West Hill              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.160497            Royal Canadian Legion       43.782533   
2              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
3              -79.188711                G & G Electronics       43.765309   
4              -79.188711                       Marina Spa       43.766000   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.189914           Pizza Place  
3       -79.191537     Electronics Store  
4       -79.191000                   Spa

Then group the venues by "Neighborhood"

In [33]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                 30   
Agincourt                                                               4   
Agincourt North,L'Amoreaux East,Milliken,Steele...                      2   
Albion Gardens,Beaumond Heights,Humbergate,Jame...                      9   
Alderwood,Long Branch                                                   9   
Bathurst Manor,Downsview North,Wilson Heights                          19   
Bayview Village                                                         4   
Bedford Park,Lawrence Manor East                                       23   
Berczy Park                                                            30   
Birch Cliff,Cliffside West                                              4   
Bloordale Gardens,Eringate,Markland Wood,Old Bu...                      7   
Brockton,Exhibition Place,Parkdale Village                             24   
Business Reply Mail Processing Centre 969 Eastern                      15   
CFB Toronto,Downsview East                                              2   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     17   
Cabbagetown,St. James Town                                             30   
Caledonia-Fairbanks                                                     5   
Canada Post Gateway Processing Centre                                  11   
Cedarbrae                                                               7   
Central Bay Street                                                     30   
Chinatown,Grange Park,Kensington Market                                30   
Christie                                                               16   
Church and Wellesley                                                   30   
Clairlea,Golden Mile,Oakridge                                           9   
Clarks Corners,Sullivan,Tam O'Shanter                                  12   
Cliffcrest,Cliffside,Scarborough Village West                           3   
Commerce Court,Victoria Hotel                                          30   
Davisville                                                             30   
Davisville North                                                        9   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     16   
...                                                                   ...   
Northwest                                                               2   
Northwood Park,York University                                          6   
Parkdale,Roncesvalles                                                  15   
Parkwoods                                                               3   
Queen's Park                                                           30   
Rosedale                                                                5   
Roselawn                                                                2   
Rouge,Malvern                                                           1   
Runnymede,Swansea                                                      30   
Ryerson,Garden District                                                30   
Scarborough Village                                                     1   
St. James Town                                                         30   
Stn A PO Boxes 25 The Esplanade                                        30   
Studio District                                                        30   
The Annex,North Midtown,Yorkville                                      22   
The Beaches                                                             6   
The Beaches West,India Bazaar                                          20   
The Danforth West,Riverdale                                            30   
The Junction North,Runnymede                                            4   
The Kingsway,Montgomery Road,Old Mill North     

We then check to see how many unique categories of venues there are in the Toronto

In [34]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 237 uniques categories.


We check to see which places are most visited by neighborhood

In [36]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

In [37]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                              Adelaide,King,Richmond     0.000000   
1                                           Agincourt     0.000000   
2   Agincourt North,L'Amoreaux East,Milliken,Steel...     0.000000   
3   Albion Gardens,Beaumond Heights,Humbergate,Jam...     0.000000   
4                               Alderwood,Long Branch     0.000000   
5       Bathurst Manor,Downsview North,Wilson Heights     0.000000   
6                                     Bayview Village     0.000000   
7                    Bedford Park,Lawrence Manor East     0.000000   
8                                         Berczy Park     0.000000   
9                          Birch Cliff,Cliffside West     0.000000   
10  Bloordale Gardens,Eringate,Markland Wood,Old B...     0.000000   
11         Brockton,Exhibition Place,Parkdale Village     0.000000   
12  Business Reply Mail Processing Centre 969 Eastern     0.066667   
13                         CFB Toronto,Downsview East     0.000000   
14  CN Tower,Bathurst Quay,Island airport,Harbourf...     0.000000   
15                         Cabbagetown,St. James Town     0.000000   
16                                Caledonia-Fairbanks     0.000000   
17              Canada Post Gateway Processing Centre     0.000000   
18                                          Cedarbrae     0.000000   
19                                 Central Bay Street     0.000000   
20            Chinatown,Grange Park,Kensington Market     0.000000   
21                                           Christie     0.000000   
22                               Church and Wellesley     0.000000   
23                      Clairlea,Golden Mile,Oakridge     0.000000   
24              Clarks Corners,Sullivan,Tam O'Shanter     0.000000   
25      Cliffcrest,Cliffside,Scarborough Village West     0.000000   
26                      Commerce Court,Victoria Hotel     0.000000   
27                                         Davisville     0.000000   
28                                   Davisville North     0.000000   
29  Deer Park,Forest Hill SE,Rathnelly,South Hill,...     0.000000   
..                                                ...          ...   
69                                          Northwest     0.000000   
70                     Northwood Park,York University     0.000000   
71                              Parkdale,Roncesvalles     0.000000   
72                                          Parkwoods     0.000000   
73                                       Queen's Park     0.033333   
74                                           Rosedale     0.000000   
75                                           Roselawn     0.000000   
76                                      Rouge,Malvern     0.000000   
77                                  Runnymede,Swansea     0.000000   
78                            Ryerson,Garden District     0.000000   
79                                Scarborough Village     0.000000   
80                                     St. James Town     0.000000   
81                    Stn A PO Boxes 25 The Esplanade     0.000000   
82                                    Studio District     0.033333   
83                  The Annex,North Midtown,Yorkville     0.000000   
84                                        The Beaches     0.000000   
85                      The Beaches West,India Bazaar     0.000000   
86                        The Danforth West,Riverdale     0.033333   
87                       The Junction North,Runnymede     0.000000   
88        The Kingsway,Montgomery Road,Old Mill North     0.000000   
89                                   Thorncliffe Park     0.058824   
90                                   Victoria Village     0.000000   
91                                          Westmount     0.000000   
92                                             Weston     0.000000   
93                                   Willowdale South     0.000000   
94                                    Willowdale We

We see the top 3 most visited venues in each neigborhood

In [38]:
num_top_venues = 3

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
        venue  freq
0        Café  0.10
1  Steakhouse  0.10
2       Hotel  0.07


----Agincourt----
            venue  freq
0          Lounge  0.25
1  Breakfast Spot  0.25
2  Clothing Store  0.25


----Agincourt North,L'Amoreaux East,Milliken,Steeles East----
           venue  freq
0     Playground   0.5
1           Park   0.5
2  Movie Theater   0.0


----Albion Gardens,Beaumond Heights,Humbergate,Jamestown,Mount Olive,Silverstone,South Steeles,Thistletown----
                  venue  freq
0         Grocery Store  0.22
1           Pizza Place  0.22
2  Fast Food Restaurant  0.11


----Alderwood,Long Branch----
          venue  freq
0   Pizza Place  0.22
1   Coffee Shop  0.11
2  Dance Studio  0.11


----Bathurst Manor,Downsview North,Wilson Heights----
         venue  freq
0  Coffee Shop  0.11
1     Pharmacy  0.05
2        Diner  0.05


----Bayview Village----
                venue  freq
0                Café  0.25
1  Chinese Restaurant  0.25
2             

Even better, we can see in a Pandas data frame the top ten common venues in each neighborhood

In [39]:
def _most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [40]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
sorted_neighborhoods_venues = pd.DataFrame(columns=columns)
sorted_neighborhoods_venues['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    sorted_neighborhoods_venues.iloc[ind, 1:] = _most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

sorted_neighborhoods_venues.head()

Neighborhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond                  Café   
1                                          Agincourt                Lounge   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...            Playground   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...         Grocery Store   
4                              Alderwood,Long Branch           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Steakhouse      Asian Restaurant                 Hotel   
1        Clothing Store        Breakfast Spot          Skating Rink   
2                  Park         Women's Store          Dance Studio   
3           Pizza Place   Fried Chicken Joint        Sandwich Place   
4                   Gym          Skating Rink           Coffee Shop   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0                Lounge                   Bar                    Speakeasy   
1      Department Store   Empanada Restaurant            Electronics Store   
2   Empanada Restaurant     Electronics Store  Eastern European Restaurant   
3  Fast Food Restaurant            Beer Store                     Pharmacy   
4                   Pub          Dance Studio               Sandwich Place   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                   Smoke Shop    Seafood Restaurant              Gastropub  
1  Eastern European Restaurant   Dumpling Restaurant              Drugstore  
2          Dumpling Restaurant             Drugstore                Dog Run  
3                      Gay Bar      Cuban Restaurant              Drugstore  
4                     Pharmacy                Garden        Coworking Space

# Clustering

In [41]:
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

kmeans.labels_[0:98]

array([0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 4, 0, 0, 0, 0, 0,
       0, 4, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 3, 0, 4, 4, 0, 4, 3, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 4, 4, 0, 0, 0, 4, 0, 0, 0,
       1, 2, 0, 0, 0, 0, 4, 0, 4, 0, 0, 0, 0, 1, 0, 0, 0, 0, 4, 0, 0, 4,
       4, 0, 0, 0, 4, 0, 0, 0, 0, 0], dtype=int32)

In [42]:
#sorted_neighborhoods_venues.drop(['Cluster Labels'],axis=1,inplace=True)
sorted_neighborhoods_venues.insert(0, 'Cluster Labels', kmeans.labels_)
toronto_merged = df_5
# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(sorted_neighborhoods_venues.set_index('Neighborhood'), on='Neighbourhood')
toronto_merged.dropna(subset=["Cluster Labels"], axis=0, inplace=True)
toronto_merged.reset_index(drop=True, inplace=True)
toronto_merged['Cluster Labels'].astype(int)
toronto_merged.head()

Postcode      Borough                         Neighbourhood   Latitude  \
0      M1B  Scarborough                         Rouge,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2      M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3      M1G  Scarborough                                Woburn  43.770992   
4      M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             0.0  Fast Food Restaurant         Women's Store   
1 -79.160497             0.0                   Bar         Women's Store   
2 -79.188711             0.0                   Spa          Intersection   
3 -79.216917             0.0           Coffee Shop     Korean Restaurant   
4 -79.239476             0.0   Fried Chicken Joint                Bakery   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0    Falafel Restaurant  Ethiopian Restaurant   Empanada Restaurant   
1        Farmers Market           Event Space  Ethiopian Restaurant   
2     Electronics Store           Pizza Place    Mexican Restaurant   
3    Falafel Restaurant  Ethiopian Restaurant   Empanada Restaurant   
4      Hakka Restaurant                  Bank       Thai Restaurant   

  6th Most Common Venue        7th Most Common Venue  \
0     Electronics Store  Eastern European Restaurant   
1   Empanada Restaurant            Electronics Store   
2        Breakfast Spot          Rental Car Location   
3     Electronics Store  Eastern European Restaurant   
4    Athletics & Sports         Caribbean Restaurant   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0          Dumpling Restaurant             Drugstore                Dog Run  
1  Eastern European Restaurant   Dumpling Restaurant              Drugstore  
2               Medical Center    Dim Sum Restaurant                  Diner  
3          Dumpling Restaurant             Drugstore                Dog Run  
4               Discount Store    Dim Sum Restaurant                  Diner

We visualize the clusters in a map

In [43]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels'].astype(int)):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)   
map_clusters

Now we examine the clusters to see the distinguishing venues

In [44]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels   1st Most Common Venue  \
0       Scarborough             0.0    Fast Food Restaurant   
1       Scarborough             0.0                     Bar   
2       Scarborough             0.0                     Spa   
3       Scarborough             0.0             Coffee Shop   
4       Scarborough             0.0     Fried Chicken Joint   
6       Scarborough             0.0             Coffee Shop   
8       Scarborough             0.0                   Motel   
9       Scarborough             0.0         College Stadium   
10      Scarborough             0.0       Indian Restaurant   
11      Scarborough             0.0             Auto Garage   
12      Scarborough             0.0                  Lounge   
13      Scarborough             0.0                Pharmacy   
15      Scarborough             0.0    Fast Food Restaurant   
16        NorthYork             0.0                 Dog Run   
17        NorthYork             0.0             Coffee Shop   
18        NorthYork             0.0      Chinese Restaurant   
20        NorthYork             0.0        Ramen Restaurant   
22        NorthYork             0.0             Pizza Place   
24        NorthYork             0.0        Basketball Court   
25        NorthYork             0.0                     Gym   
26        NorthYork             0.0             Coffee Shop   
27        NorthYork             0.0             Coffee Shop   
31        NorthYork             0.0           Grocery Store   
32        NorthYork             0.0             Coffee Shop   
33         EastYork             0.0             Pizza Place   
34         EastYork             0.0                    Park   
36         EastYork             0.0     Sporting Goods Shop   
37         EastYork             0.0            Burger Joint   
39      EastToronto             0.0        Greek Restaurant   
40      EastToronto             0.0                    Park   
..              ...             ...                     ...   
60        NorthYork             0.0             Coffee Shop   
61   CentralToronto             0.0                  Garden   
62   CentralToronto             0.0           Jewelry Store   
63   CentralToronto             0.0                    Café   
64  DowntownToronto             0.0                    Café   
65  DowntownToronto             0.0                    Café   
66  DowntownToronto             0.0         Airport Service   
67  DowntownToronto             0.0                    Café   
68  DowntownToronto             0.0                    Café   
69        NorthYork             0.0  Furniture / Home Store   
70        NorthYork             0.0             Pizza Place   
73  DowntownToronto             0.0           Grocery Store   
74      WestToronto             0.0             Supermarket   
75      WestToronto             0.0                     Bar   
76      WestToronto             0.0                    Café   
78             York             0.0              Restaurant   
80      WestToronto             0.0         Thai Restaurant   
81      WestToronto             0.0          Breakfast Spot   
82      WestToronto             0.0                    Café   
83      Queen'sPark             0.0             Coffee Shop   
84      Mississauga             0.0             Coffee Shop   
85      EastToronto             0.0             Yoga Studio   
86        Etobicoke             0.0                    Café   
87        Etobicoke             0.0             Pizza Place   
90        Etobicoke             0.0                     Gym   
91        Etobicoke             0.0             Pizza Place   
92        NorthYork             0.0             Pizza Place   
95        Etobicoke             0.0             Pizza Place   
97        Etobicoke             0.0           Grocery Store   
98        Etobicoke             0.0               Drugstore   

        2nd Most Common Venue  3rd Most Common Venue  \
0               Women's Store     Falafel Restaur

In [45]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
5      Scarborough             1.0            Playground   
14     Scarborough             1.0            Playground   
46  CentralToronto             1.0          Tennis Court   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5          Women's Store          Dance Studio   Empanada Restaurant   
14                  Park         Women's Store          Dance Studio   
46            Playground           Summer Camp         Women's Store   

   5th Most Common Venue        6th Most Common Venue  \
5      Electronics Store  Eastern European Restaurant   
14   Empanada Restaurant            Electronics Store   
46     Electronics Store  Eastern European Restaurant   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
5           Dumpling Restaurant             Drugstore               Dog Run   
14  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
46          Dumpling Restaurant             Drugstore               Dog Run   

   10th Most Common Venue  
5          Discount Store  
14                Dog Run  
46         Discount Store

In [46]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
19  NorthYork             2.0             Piano Bar          Dance Studio   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
19   Empanada Restaurant     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
19   Dumpling Restaurant             Drugstore               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
19        Discount Store                  Diner

In [47]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
30  NorthYork             3.0     Korean Restaurant        Baseball Field   
93  NorthYork             3.0        Baseball Field         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
30            Food Truck         Women's Store          Dessert Shop   
93        Farmers Market           Event Space  Ethiopian Restaurant   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
30  Ethiopian Restaurant   Empanada Restaurant            Electronics Store   
93   Empanada Restaurant     Electronics Store  Eastern European Restaurant   

          9th Most Common Venue 10th Most Common Venue  
30  Eastern European Restaurant    Dumpling Restaurant  
93          Dumpling Restaurant              Drugstore

In [48]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
7       Scarborough             4.0                Bakery   
21        NorthYork             4.0                  Park   
23        NorthYork             4.0                  Park   
28        NorthYork             4.0               Airport   
29        NorthYork             4.0         Grocery Store   
35      EastToronto             4.0                 Trail   
38         EastYork             4.0           Coffee Shop   
42   CentralToronto             4.0                  Park   
43   CentralToronto             4.0                 Hotel   
48  DowntownToronto             4.0                  Park   
71             York             4.0                 Trail   
72             York             4.0                  Park   
77        NorthYork             4.0                  Park   
79             York             4.0           Pizza Place   
88        Etobicoke             4.0                  Park   
89        Etobicoke             4.0      Business Service   
94             York             4.0     Convenience Store   
96        Etobicoke             4.0           Pizza Place   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
7                     Bus Line           Bus Station                  Park   
21           Convenience Store                  Bank         Women's Store   
23           Food & Drink Shop             BBQ Joint         Women's Store   
28                        Park         Women's Store         Deli / Bodega   
29                        Park         Shopping Mall                  Bank   
35                        Park      Asian Restaurant     Health Food Store   
38                        Park     Convenience Store         Deli / Bodega   
42                 Swim School              Bus Line         Women's Store   
43                         Gym        Breakfast Spot     Food & Drink Shop   
48                       Trail            Playground              Building   
71                        Park                 Field          Hockey Arena   
72               Women's Store  Fast Food Restaurant                Market   
77  Construction & Landscaping      Basketball Court                Bakery   
79                    Bus Line     Convenience Store         Grocery Store   
88                  Smoke Shop                 River         Women's Store   
89                        Park        Baseball Field         Deli / Bodega   
94               Women's Store         Deli / Bodega  Ethiopian Restaurant   
96                        Park              Bus Line     Mobile Phone Shop   

     5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
7             Soccer Field  Fast Food Restaurant                 Intersection   
21        Department Store  Ethiopian Restaurant          Empanada Restaurant   
23        Department Store   Empanada Restaurant            Electronics Store   
28    Ethiopian Restaurant   Empanada Restaurant            Electronics Store   
29           Women's Store      Department Store          Empanada Restaurant   
35                     Pub         Women's Store                 Dessert Shop   
38     Empanada Restaurant     Electronics Store  Eastern European Restaurant   
42        Department Store   Empanada Restaurant            Electronics Store   
43       Convenience Store        Sandwich Place                         Park   
48           Women's Store      Department Store                 Dessert Shop   
71           Women's Store      Department Store                 Dessert Shop   
72        Department Store   Empanada Restaurant            Electronics Store   
77           Women's Store          Dessert Shop         Ethiopian Restaurant   
79  Gluten-free Restaurant             Gift Shop  Eastern European Restaurant   
88            Dance Studio     Electronics Store  Eastern European Restaurant   
89     Empanada Restaurant     Electronics Store  Eastern European Restau

The segmentation and clustering is achieved